File: src/main/java/com/lessmarkup/engine/data/migrate/MigratorImpl.java
/*
 * This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0/.
 */
package com.lessmarkup.engine.data.migrate;

import com.lessmarkup.Constants;
import com.lessmarkup.engine.data.ConnectionManager;
import com.lessmarkup.engine.data.dialects.DatabaseLanguageDialect;
import com.lessmarkup.engine.data.dialects.DatabaseDataType;
import com.lessmarkup.engine.data.dialects.DatabaseLanguageDialectFactory;
import com.lessmarkup.framework.helpers.LoggingHelper;
import com.lessmarkup.framework.helpers.PropertyDescriptor;
import com.lessmarkup.framework.helpers.StringHelper;
import com.lessmarkup.framework.helpers.TypeHelper;
import com.lessmarkup.interfaces.data.DataObject;
import com.lessmarkup.interfaces.data.Migrator;
import com.lessmarkup.interfaces.data.OptionalBoolean;
import com.lessmarkup.interfaces.exceptions.CommonException;
import com.lessmarkup.interfaces.exceptions.DatabaseException;

import java.sql.Connection;
import java.sql.ResultSet;
import java.sql.SQLException;
import java.sql.Statement;
import java.time.OffsetDateTime;
import java.util.*;

import com.lessmarkup.interfaces.recordmodel.MaxLength;
import com.lessmarkup.interfaces.recordmodel.RequiredField;
import org.atteo.evo.inflector.English;

public class MigratorImpl implements Migrator {

    private final Connection connection;
    private final DatabaseLanguageDialect dialect;
    
    public MigratorImpl(String connectionString) {
        if (StringHelper.isNullOrEmpty(connectionString)) {
            this.connection = null;
            this.dialect = null;
            return;
        }
        try {
            this.connection = ConnectionManager.getConnection(connectionString);
            this.dialect = DatabaseLanguageDialectFactory.createDialect(this.connection);
        } catch (SQLException e) {
            throw new DatabaseException(e);
        } catch (ClassNotFoundException ex) {
            throw new CommonException(ex);
        }
    }
    
    @Override
    public void executeSql(String sql) {
        if (connection == null) {
            return;
        }
        try (Statement statement = this.connection.createStatement()) {
            statement.execute(sql);
        } catch (SQLException e) {
            throw new DatabaseException(e);
        }
    }
    
    public Object executeScalar(String sql) {
        if (connection == null) {
            return null;
        }
        try (Statement statement = this.connection.createStatement();
                ResultSet resultSet = statement.executeQuery(sql)) {
            if (!resultSet.first()) {
                return null;
            }
            return resultSet.getObject(1);
        } catch (SQLException e) {
            throw new DatabaseException(e);
        }
    }
    
    @Override
    public <T extends DataObject> boolean checkExists(Class<T> type) {
        if (connection == null) {
            return true;
        }
        String tableName = English.plural(type.getSimpleName());
        return (long) executeScalar(String.format("SELECT COUNT(*) FROM INFORMATION_SCHEMA.TABLES WHERE TABLE_NAME = '%s'", tableName)) > 0;
    }
    
    private String getDataType(PropertyDescriptor property) {

        Class<?> propertyType = property.getType();
        
        if (propertyType.equals(OptionalLong.class)) {
            return this.dialect.getTypeDeclaration(DatabaseDataType.LONG(), false);
        }
        
        if (propertyType.equals(OptionalInt.class)) {
            return this.dialect.getTypeDeclaration(DatabaseDataType.INT(), false);
        }

        if (propertyType.equals(OptionalBoolean.class)) {
            return this.dialect.getTypeDeclaration(DatabaseDataType.BOOLEAN(), false);
        }

        if (propertyType.equals(OptionalDouble.class)) {
            return this.dialect.getTypeDeclaration(DatabaseDataType.DOUBLE(), false);
        }

        boolean required = property.getAnnotation(RequiredField.class) != null;
        OptionalInt maxLength = OptionalInt.empty();

        MaxLength maxLengthAnnotation = property.getAnnotation(MaxLength.class);
        if (maxLengthAnnotation != null) {
            maxLength = OptionalInt.of(maxLengthAnnotation.length());
        }

        if (propertyType.equals(int.class) || propertyType.equals(Integer.class)) {
            return this.dialect.getTypeDeclaration(DatabaseDataType.INT(), required);
        }
        
        if (propertyType.equals(OffsetDateTime.class)) {
            return this.dialect.getTypeDeclaration(DatabaseDataType.DATE_TIME(), required);
        }

        if (propertyType.equals(long.class) || propertyType.equals(Long.class)) {
            if (property.getName().equals(Constants.Data.ID_PROPERTY_NAME)) {
                return this.dialect.getTypeDeclaration(DatabaseDataType.IDENTITY(), required);
            }
            return this.dialect.getTypeDeclaration(DatabaseDataType.LONG(), required);
        }

        if (propertyType.equals(boolean.class) || propertyType.equals(Boolean.class)) {
            return this.dialect.getTypeDeclaration(DatabaseDataType.BOOLEAN(), required);
        }

        if (propertyType.equals(double.class) || propertyType.equals(Double.class)) {
            return this.dialect.getTypeDeclaration(DatabaseDataType.DOUBLE(), required);
        }

        if (propertyType.equals(String.class)) {
            return this.dialect.getTypeDeclaration(DatabaseDataType.STRING(), maxLength, required, null);
        }

        if (propertyType.equals(byte[].class)) {
            return this.dialect.getTypeDeclaration(DatabaseDataType.BINARY(), maxLength, required, null);
        }

        if (propertyType.isEnum()) {
            return this.dialect.getTypeDeclaration(DatabaseDataType.INT(), required);
        }

        throw new IllegalArgumentException();
    }

    @Override
    public <T extends DataObject> void createTable(Class<T> type) {
        if (checkExists(type)) {
            updateTable(type);
            return;
        }

        String tableName = English.plural(type.getSimpleName());

        final StringBuilder sb = new StringBuilder();
        sb.append(String.format("CREATE TABLE %s (", this.dialect.decorateName(tableName)));

        boolean first = true;
        
        for (PropertyDescriptor property : TypeHelper.getProperties(type)) {
            if (first) {
                first = false;
            } else {
                sb.append(", ");
            }
            String dataType = getDataType(property);
            sb.append(String.format("%s %s", this.dialect.decorateName(property.getName()), dataType));
        }
        
        sb.append(String.format(", CONSTRAINT %s PRIMARY KEY (%s ASC))", this.dialect.decorateName("PK_" + tableName), this.dialect.decorateName(Constants.Data.ID_PROPERTY_NAME)));

        try {
            executeSql(sb.toString());
        } catch (Exception e) {
            LoggingHelper.getLogger(getClass()).info("Error creating table: " + sb.toString());
            throw e;
        }
    }

    <TD extends DataObject, TB extends DataObject> boolean checkDependency(Class<TD> typeD, Class<TB> typeB, String column) {
        if (connection == null) {
            return true;
        }
        
        String dependentTableName = English.plural(typeD.getSimpleName());
        String baseTableName = English.plural(typeB.getSimpleName());

        if (column == null) {
            column = String.format("%sId", typeB.getSimpleName());
        }

        String text = String.format("SELECT COUNT(*) FROM INFORMATION_SCHEMA.REFERENTIAL_CONSTRAINTS WHERE CONSTRAINT_NAME = 'FK_%s_%s_%s'", dependentTableName, baseTableName, column);

        return (long) executeScalar(text) > 0;
    }
    
    @Override
    public <TD extends DataObject, TB extends DataObject> void addDependency(Class<TD> typeD, Class<TB> typeB, String column) {
        if (connection == null) {
            return;
        }
        
        if (checkDependency(typeD, typeB, column)) {
            return;
        }

        String dependentTableName = English.plural(typeD.getSimpleName());
        String baseTableName = English.plural(typeB.getSimpleName());

        if (column == null) {
            column = String.format("%sId", typeB.getSimpleName());
        }

        String text = String.format("ALTER TABLE %s ADD CONSTRAINT %s FOREIGN KEY(%s) REFERENCES %s (%s)",
                this.dialect.decorateName(dependentTableName),
                this.dialect.decorateName(String.format("FK_%s_%s_%s", dependentTableName, baseTableName, column)),
                this.dialect.decorateName(column),
                this.dialect.decorateName(baseTableName),
                this.dialect.decorateName(Constants.Data.ID_PROPERTY_NAME));

        executeSql(text);

        text = String.format("CREATE INDEX %s ON %S (%s ASC)",
                this.dialect.decorateName("IX_" + column),
                this.dialect.decorateName(dependentTableName),
                this.dialect.decorateName(column));

        executeSql(text);
    }

    @Override
    public <TD extends DataObject, TB extends DataObject> void deleteDependency(Class<TD> typeD, Class<TB> typeB, String column) {
        if (connection == null) {
            return;
        }

        final String dependentTableName = English.plural(typeD.getSimpleName());
        final String baseTableName = English.plural(typeB.getSimpleName());

        if (column == null) {
            column = String.format("%sId", typeB.getSimpleName());
        }

        String text = String.format("DROP INDEX %s ON %s", this.dialect.decorateName("IX_" + column), this.dialect.decorateName(dependentTableName));

        executeSql(text);

        text = String.format("ALTER TABLE %s DROP CONSTRAINT %s",
                this.dialect.decorateName(dependentTableName),
                this.dialect.decorateName(String.format("FK_%s_%s_%s", dependentTableName, baseTableName, column)));

        executeSql(text);
    }

    @Override
    public <T extends DataObject> void updateTable(Class<T> type) {
        if (connection == null) {
            return;
        }

        final String tableName = English.plural(type.getSimpleName());
        
        final Map<String, String> columnsToAdd = new HashMap<>();
        
        final String columnId = Constants.Data.ID_PROPERTY_NAME.toLowerCase();
        
        for (PropertyDescriptor property : TypeHelper.getProperties(type)) {
            String columnName = property.getName().toLowerCase();
            if (columnName.equals(columnId)) {
                continue;
            }
            columnsToAdd.put(columnName, getDataType(property));
        }

        List<String> columnsToDrop = new ArrayList<>();

        try (Statement statement = this.connection.createStatement()) {
            try (ResultSet result = statement.executeQuery(
                    "SELECT COLUMN_NAME, IS_NULLABLE, DATA_TYPE, CHARACTER_MAXIMUM_LENGTH, CHARACTER_SET_NAME " + 
                    "FROM INFORMATION_SCHEMA.COLUMNS WHERE TABLE_NAME = '" + tableName + "'")) {
                while (result.next()) {
                    String columnName = result.getString(1);
                    String columnNameLower = columnName.toLowerCase();

                    if (columnId.equals(columnNameLower)) {
                        continue;
                    }

                    final String currentType = columnsToAdd.get(columnNameLower);
                    if (currentType == null) {
                        columnsToDrop.add(columnName);
                        continue;
                    }

                    final boolean required = !"YES".equals(result.getString(2));
                    final DatabaseDataType dataType = this.dialect.getDataType(result.getString(3));
                    if (dataType == null) {
                        columnsToDrop.add(columnName);
                        continue;
                    }
                    OptionalInt sizeLimit = OptionalInt.empty();
                    int value = result.getInt(4);
                    if (!result.wasNull() && value != 65535) {
                        sizeLimit = OptionalInt.of(value);
                    }
                    
                    String characterSet = result.getString(5);
                    if (result.wasNull()) {
                        characterSet = null;
                    }

                    final String columnType = this.dialect.getTypeDeclaration(dataType, sizeLimit, required, characterSet);

                    if (!currentType.equals(columnType)) {
                        columnsToDrop.add(columnName);
                    }
                    else {
                        columnsToAdd.remove(columnNameLower);
                    }
                }
            }
        } catch (SQLException e) {
            throw new DatabaseException(e);
        }
        
        for (String column : columnsToDrop) {
            executeSql(String.format("ALTER TABLE %s DROP COLUMN %s", this.dialect.decorateName(tableName), this.dialect.decorateName(column)));
        }
        
        for (Map.Entry<String, String> entry : columnsToAdd.entrySet()) {
            executeSql(String.format("ALTER TABLE %s ADD %s %s", this.dialect.decorateName(tableName), this.dialect.decorateName(entry.getKey()), entry.getValue()));
        }
    }

    @Override
    public <T extends DataObject> void deleteTable(Class<T> type) {
        if (connection == null) {
            return;
        }
        String text = String.format("DROP TABLE %s", this.dialect.decorateName(English.plural(type.getSimpleName())));
        executeSql(text);
    }
}

File: src/main/scala/com/lessmarkup/engine/data/migrate/MigratorImpl.scala
/*
 * This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0/.
 */

package com.lessmarkup.engine.data.migrate

import java.sql.{Connection, ResultSet, SQLException, Statement}
import java.time.OffsetDateTime
import com.lessmarkup.Constants
import com.lessmarkup.engine.data.ConnectionManager
import com.lessmarkup.engine.data.dialects.{DatabaseDataType, DatabaseLanguageDialect, DatabaseLanguageDialectFactory}
import com.lessmarkup.framework.helpers.{LoggingHelper, PropertyDescriptor, StringHelper, TypeHelper}
import com.lessmarkup.interfaces.annotations.{RequiredField, MaxLength}
import com.lessmarkup.interfaces.data._
import com.lessmarkup.interfaces.exceptions.DatabaseException
import org.atteo.evo.inflector.English

import scala.collection.JavaConversions._

class MigratorImpl(connectionString: String) extends Migrator {

  private val connection: Option[Connection] =
    if (StringHelper.isNullOrEmpty(connectionString))
      None
    else
      Option(ConnectionManager.getConnection(connectionString))

  private val dialect: Option[DatabaseLanguageDialect] =
    if (connection.isDefined)
      Option(DatabaseLanguageDialectFactory.createDialect(connection))
    else
      None

  def executeSql(sql: String) {

    if (connection.isEmpty) {
      return
    }

    val statement: Statement = connection.get.createStatement
    try {
      statement.execute(sql)
    }
    catch {
      case e: SQLException => throw new DatabaseException(e)
    } finally {
      if (statement != null) statement.close()
    }
  }

  def executeScalar(sql: String): AnyRef = {

    if (connection.isEmpty) {
      return null
    }

    val statement: Statement = this.connection.get.createStatement
    val resultSet: ResultSet = statement.executeQuery(sql)
    try {
      if (!resultSet.first) {
        return null
      }
      resultSet.getObject(1)
    }
    catch {
      case e: SQLException => throw new DatabaseException(e)
    } finally {
      if (statement != null) statement.close()
      if (resultSet != null) resultSet.close()
    }
  }

  def checkExists[T <: DataObject](`type`: Class[T]): Boolean = {
    if (connection == null) {
      return true
    }
    val tableName: String = English.plural(`type`.getSimpleName)
    executeScalar(s"SELECT COUNT(*) FROM INFORMATION_SCHEMA.TABLES WHERE TABLE_NAME = '$tableName'").asInstanceOf[Long] > 0
  }

  private def getDataType(property: PropertyDescriptor): String = {

    if (dialect.isEmpty) {
      throw new IllegalArgumentException
    }

    val propertyType: Class[_] = property.getType
    if (propertyType == classOf[OptionLong]) {
      return dialect.get.getTypeDeclaration(DatabaseDataType.LONG, required = false)
    }
    if (propertyType == classOf[OptionInt]) {
      return dialect.get.getTypeDeclaration(DatabaseDataType.INT, required = false)
    }
    if (propertyType == classOf[OptionBool]) {
      return dialect.get.getTypeDeclaration(DatabaseDataType.BOOLEAN, required = false)
    }
    if (propertyType == classOf[OptionDouble]) {
      return dialect.get.getTypeDeclaration(DatabaseDataType.DOUBLE, required = false)
    }
    val required: Boolean = property.getAnnotation(classOf[RequiredField]) != null
    val maxLengthAnnotation: MaxLength = property.getAnnotation(classOf[MaxLength])
    val maxLength: Option[Int] =
      if (maxLengthAnnotation != null) Option(maxLengthAnnotation.length)
      else None
    if ((propertyType == classOf[Int]) || (propertyType == classOf[Integer])) {
      return dialect.get.getTypeDeclaration(DatabaseDataType.INT, required)
    }
    if (propertyType == classOf[OffsetDateTime]) {
      return dialect.get.getTypeDeclaration(DatabaseDataType.DATE_TIME, required)
    }
    if ((propertyType == classOf[Long]) || (propertyType == classOf[Long])) {
      if (property.getName == Constants.DataIdPropertyName) {
        return dialect.get.getTypeDeclaration(DatabaseDataType.IDENTITY, required)
      }
      return dialect.get.getTypeDeclaration(DatabaseDataType.LONG, required)
    }
    if ((propertyType == classOf[Boolean]) || (propertyType == classOf[Boolean])) {
      return dialect.get.getTypeDeclaration(DatabaseDataType.BOOLEAN, required)
    }
    if ((propertyType == classOf[Double]) || (propertyType == classOf[Double])) {
      return dialect.get.getTypeDeclaration(DatabaseDataType.DOUBLE, required)
    }
    if (propertyType == classOf[String]) {
      return dialect.get.getTypeDeclaration(DatabaseDataType.STRING, maxLength, required, null)
    }
    if (propertyType == classOf[Array[Byte]]) {
      return dialect.get.getTypeDeclaration(DatabaseDataType.BINARY, maxLength, required, null)
    }
    if (propertyType.isEnum) {
      return dialect.get.getTypeDeclaration(DatabaseDataType.INT, required)
    }
    throw new IllegalArgumentException
  }

  def createTable[T <: DataObject](tableType: Class[T]) {

    if (dialect.isEmpty) {
      return
    }

    if (checkExists(tableType)) {
      updateTable(tableType)
      return
    }

    val tableName: String = English.plural(tableType.getSimpleName)
    val sb: StringBuilder = new StringBuilder
    sb.append(s"CREATE TABLE ${dialect.get.decorateName(tableName)} (")

    sb.append(TypeHelper.getProperties(tableType).map(t => {
      val dataType: String = getDataType(t)
      String.format("%s %s", dialect.get.decorateName(t.getName), dataType)
    }).mkString(", "))

    sb.append(String.format(", CONSTRAINT %s PRIMARY KEY (%s ASC))",
      dialect.get.decorateName("PK_" + tableName),
      dialect.get.decorateName(Constants.DataIdPropertyName)))

    try {
      executeSql(sb.toString())
    }
    catch {
      case e: Exception =>
        LoggingHelper.getLogger(getClass).info("Error creating table: " + sb.toString)
        throw e
    }
  }

  private[migrate] def checkDependency[TD <: DataObject, TB <: DataObject](typeD: Class[TD], typeB: Class[TB], column: String): Boolean = {

    if (connection.isEmpty) {
      return true
    }
    val dependentTableName: String = English.plural(typeD.getSimpleName)
    val baseTableName: String = English.plural(typeB.getSimpleName)

    val safeColumn = if (column == null) String.format("%sId", typeB.getSimpleName) else column

    val text: String = String.format("SELECT COUNT(*) FROM INFORMATION_SCHEMA.REFERENTIAL_CONSTRAINTS WHERE CONSTRAINT_NAME = 'FK_%s_%s_%s'",
      dependentTableName, baseTableName, safeColumn)

    executeScalar(text).asInstanceOf[Long] > 0
  }

  def addDependency[TD <: DataObject, TB <: DataObject](typeD: Class[TD], typeB: Class[TB], column: String) {

    if (connection.isEmpty) {
      return
    }

    if (checkDependency(typeD, typeB, column)) {
      return
    }

    val dependentTableName: String = English.plural(typeD.getSimpleName)
    val baseTableName: String = English.plural(typeB.getSimpleName)
    val safeColumn = if (column == null) String.format("%sId", typeB.getSimpleName) else column

    val command1: String = String.format("ALTER TABLE %s ADD CONSTRAINT %s FOREIGN KEY(%s) REFERENCES %s (%s)",
      dialect.get.decorateName(dependentTableName),
      dialect.get.decorateName(String.format("FK_%s_%s_%s", dependentTableName, baseTableName, safeColumn)),
      dialect.get.decorateName(safeColumn),
      dialect.get.decorateName(baseTableName),
      dialect.get.decorateName(Constants.DataIdPropertyName))

    executeSql(command1)

    val command2 = String.format("CREATE INDEX %s ON %S (%s ASC)",
      dialect.get.decorateName("IX_" + column),
      dialect.get.decorateName(dependentTableName),
      dialect.get.decorateName(column))
    executeSql(command2)
  }

  def deleteDependency[TD <: DataObject, TB <: DataObject](typeD: Class[TD], typeB: Class[TB], column: String) {

    if (connection.isEmpty) {
      return
    }

    val dependentTableName: String = English.plural(typeD.getSimpleName)
    val baseTableName: String = English.plural(typeB.getSimpleName)
    val safeColumn = if (column == null) String.format("%sId", typeB.getSimpleName) else column

    val command1: String = String.format("DROP INDEX %s ON %s",
      dialect.get.decorateName("IX_" + safeColumn),
      dialect.get.decorateName(dependentTableName))
    executeSql(command1)
    val command2 = String.format("ALTER TABLE %s DROP CONSTRAINT %s",
      dialect.get.decorateName(dependentTableName),
      dialect.get.decorateName(String.format("FK_%s_%s_%s", dependentTableName, baseTableName, safeColumn)))
    executeSql(command2)
  }

  class UpdatingColumn(val name: String,
                       val nameLower: String,
                       val dataType: Option[DatabaseDataType],
                       val required: Boolean,
                       val sizeLimit: Option[Int],
                       val characterSet: Option[String],
                       val columnType: String,
                       val currentType: Option[String] )

  def getColumnToUpdate(result: ResultSet, columnsToAdd: Map[String, String]): UpdatingColumn = {
    val columnName: String = result.getString(1)
    val columnNameLower = columnName.toLowerCase
    val required: Boolean = !("YES" == result.getString(2))
    val dataType: DatabaseDataType = dialect.get.getDataType(result.getString(3))
    val intValue: Int = result.getInt(4)
    val sizeLimit = if (!result.wasNull && intValue != 65535) Option(intValue) else None
    val charValue = result.getString(5)
    val characterSet = if (result.wasNull) None else Option(charValue)
    val columnType: String = dialect.get.getTypeDeclaration(dataType, sizeLimit, required, characterSet)
    new UpdatingColumn(columnName, columnNameLower, Option(dataType), required, sizeLimit, characterSet, columnType, columnsToAdd.get(columnNameLower))
  }

  def getColumnsToUpdate(tableName: String, columnsToAdd: Map[String, String]) = {
    val statement: Statement = connection.get.createStatement
    try {
      val result: ResultSet = statement.executeQuery(
        "SELECT COLUMN_NAME, IS_NULLABLE, DATA_TYPE, CHARACTER_MAXIMUM_LENGTH, CHARACTER_SET_NAME " +
          "FROM INFORMATION_SCHEMA.COLUMNS WHERE TABLE_NAME = '" + tableName + "'")
      try {
        Stream.continually(if (result.next) Option(getColumnToUpdate(result, columnsToAdd)) else None)
          .takeWhile(_.isDefined)
          .map(_.get)
      } finally {
        if (result != null) result.close()
      }
    } finally {
      if (statement != null) statement.close()
    }
  }

  def updateTable[T <: DataObject](`type`: Class[T]) {

    if (connection.isEmpty) {
      return
    }

    val tableName: String = English.plural(`type`.getSimpleName)
    val columnId: String = Constants.DataIdPropertyName.toLowerCase

    val columnsToAdd = TypeHelper.getProperties(`type`).map(p => (p, p.getName.toLowerCase)).filter(_._2 != columnId).map(p => {
      (p._2, getDataType(p._1))
    }).toMap

    val existingColumns = getColumnsToUpdate(tableName, columnsToAdd)

    val columnsToDrop = existingColumns.filter(c => c.nameLower == columnId
      || c.currentType.isEmpty
      || c.dataType.isEmpty
      || c.currentType.get != c.columnType).map(_.name).toSet

    val columnsToLeave = existingColumns.filter(c => !columnsToDrop.contains(c.name)).map(_.name).toSet

    for (column <- columnsToDrop) {
      executeSql(String.format("ALTER TABLE %s DROP COLUMN %s",
        dialect.get.decorateName(tableName),
        dialect.get.decorateName(column)))
    }

    for (entry <- columnsToAdd.filter(c => !columnsToLeave.contains(c._1)).entrySet) {
      executeSql(String.format("ALTER TABLE %s ADD %s %s",
        dialect.get.decorateName(tableName),
        dialect.get.decorateName(entry.getKey), entry.getValue))
    }
  }

  def deleteTable[T <: DataObject](`type`: Class[T]) {
    if (connection == null) {
      return
    }
    val text: String = String.format("DROP TABLE %s", dialect.get.decorateName(English.plural(`type`.getSimpleName)))
    executeSql(text)
  }
}
